[GEN][ZH] Add endian compat for BIGFileSystems

Mauller opened the pull request:
This PR is a prerequisit to a PR to unify AsciiString, it adds endian compatibility functions within an endian_compat header.

These are then used in the BIGFileSystems as they previously reloed on the networking ntohl function.
For StdBIGFileSystem this was being indirectly provided through the windows header in AsciiString. And directly included in win32BigFileSystem through winsock2.

This change is necessary to provide cross compat byte swapping functionality.

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/StdDevice/Common/StdBIGFileSystem.cpp:
```diff
@@ -113,7 +114,7 @@ ArchiveFile * StdBIGFileSystem::openArchiveFile(const Char *filename) {
 	// change the order of the bytes cause the file size is in reverse byte order for some reason.
 	fp->read(&numLittleFiles, 4);
-	numLittleFiles = ntohl(numLittleFiles);
+	numLittleFiles = bswap32(numLittleFiles);
```
> Built-in Function: uint16_t __builtin_bswap16 (uint16_t x)
Returns x with the order of the bytes reversed; for example, 0xabcd becomes 0xcdab. Byte here always means exactly 8 bits.

https://gcc.gnu.org/onlinedocs/gcc/Byte-Swapping-Builtins.html

I do not think this is the intent here. This should only swap one of Big Endian or Little Endian, no?

Mauller replied:
ntohl is byte swapping 32bit from network to hardware byte ordering, Big -> little endian.

but the builtin byteswap intrinsics can be used to genericaly byte swap regardless of system endianess, most modern systems are little endian anyway, so checking if the local system is already big endian is a moot point overall.

Mauller replied:
It also looks like the windows ntohl functions doesn't do any byte order checking anyway and just byte swaps regardless.

https://learn.microsoft.com/en-us/windows/win32/api/winsock/nf-winsock-ntohl

xezon replied:
For reference, Thyme uses this:
https://github.com/TheAssemblyArmada/Thyme/blob/ccef1e11c1355c6db577a057c06e7d790f1a0333/deps/baseconfig/src/endiantype.h

Mauller replied:
> For reference, Thyme uses this: https://github.com/TheAssemblyArmada/Thyme/blob/ccef1e11c1355c6db577a057c06e7d790f1a0333/deps/baseconfig/src/endiantype.h

That's cool, i guess it depends on how far we want to go with this.

The big archives themselves will be compressed in little endian format for the data, so the big file system as it is doesn't support big endian systems anyway. The only places we really need to worry about maintaining specific endianess are network facing functions.

My consideration was to just have some generic byte swapping functions for where they are required. And when we have code that touches network stuff, we can use system ntohl type functions there.

xezon replied:
I suggest OmniBlade chimes in on this. My understanding is that byte swap makes no sense if that just swaps on both BE and LE systems. The point of endian byte swap is so that systems correctly read in the binary data depending on whether they need swapping.

Mauller replied:
I guess a biger question at this point would be if we also want to support big endian systems?

It would require far more extensive changes in the game code to support them as a lot of the game data formats are little endian so the systems using them inherently consider the data to be little endian. 

A lot of modern systems that the majority of people would play the game on are little endian as well. Big endian systems are not very common. 

X86, Arm, Risc-v etc are all little endian and modern PowerPc can run both. Typically running in little endian nowadays.

xezon replied:
OmniBlade did build up Thyme with endian correctness.

JohnsterID replied:
For awareness, if I'm reading this conversation correctly, BFME also uses BIG4 (Big Endian).
https://web.archive.org/web/20230818234324/http://wiki.xentax.com/index.php/EA_VIV_BIG4
https://web.archive.org/web/20230818024543/http://wiki.xentax.com/index.php/EA_BIG_BIGF_Archive

xezon commented on Dependencies/Utility/Utility/endian_compat.h:
I would also like we use the htobe, htole, betoh, letoh from https://github.com/TheAssemblyArmada/Thyme/blob/develop/deps/baseconfig/src/endiantype.h

Reason being, it is bonkers to always specify the bit width in the function call explicitly (C style). It could lead to errors perhaps. The template functions automatically choose the correct width from the passed value type.

Mauller replied:
I am very used to being explicit with the calls becasue of the type of work i do, but will look at adding the helper functions.

xezon commented on Dependencies/Utility/Utility/endian_compat.h:
```diff
@@ -0,0 +1,173 @@
+//Endian helper functions
+static_assert(sizeof(SwapType16) == 2, "expected size does not match");
+static_assert(sizeof(SwapType32) == 4, "expected size does not match");
+static_assert(sizeof(SwapType64) == 8, "expected size does not match");
```
This will not compile in VS6. Could add a STATIC_ASSERT macro if necessary.

Mauller replied:
Added a preprocessor guard in the header that should assert an error during compilation.

xezon commented on GeneralsMD/Code/GameEngineDevice/Source/StdDevice/Common/StdBIGFileSystem.cpp:
```diff
@@ -113,7 +114,7 @@ ArchiveFile * StdBIGFileSystem::openArchiveFile(const Char *filename) {
 	// change the order of the bytes cause the file size is in reverse byte order for some reason.
 	fp->read(&numLittleFiles, 4);
-	numLittleFiles = ntohl(numLittleFiles);
+	numLittleFiles = betoh(numLittleFiles);
```
What about the ntohl calls in `Win32BIGFileSystem::openArchiveFile`. Shouldn't we adjust it there as well?

xezon replied:
Can't we just do `#define be32toh(x) ntohl(x)` for VS6?

Mauller replied:
I would have to add a define around all the incompatible stuff in the endian_compat header and remove the compile guard i just added at the beggining of the header if we wanted to use it with the current Win32BigFilesystem and VC6.

Will have a look at it again tomorrow after work.

xezon replied:
I think it is also fine if STATIC_ASSERT simply does nothing in VS6.

Mauller replied:
Yeah i guess the VS22 and other builds supporting static_assert should pickup on the problem and then be correct for the VC6 build.

```CPP
#if defined(_MSC_VER) && (_MSC_VER < 1300)
#define STATIC_ASSERT(expr, msg)
#else
#define STATIC_ASSERT(expr, msg) static_assert(expr, msg)
#endif
```

Should do in the cpp macros header if we wanted to more extensively use static_assert.

xezon replied:
The second one, because then we do not need to write it in upper case.

Mauller commented:
Updated, Rebased off main and merge comflict fixed.

Tweaked with discussed changes and added endian compat functions to win32Bigfilesystems for generals and zero hour along with stdbigfilesystem.

Should be good to go now.

Mauller commented:
> VC6 does not like the variadic macro. Then we need to do `#define static_assert(a,b)` and we settle on static asserts always requiring 2 arguments for now.

Yeah i just noticed this, was just about to change it, should be up in a few seconds.

Mauller commented:
Now it doesn't like long long, this is taking a long long time. i should really sort my local VC6 build environment lol.

xezon reviewed (approved):
Looks good.

Mauller commented:
> Looks good.

It would seem that VC6 doesn't like the helper templates

xezon commented:
It is using what is called "partial template class specialization". Perhaps VC6 does not support that?

Mauller commented:
> It is using what is called "partial template class specialization". Perhaps VC6 does not support that?

I have not used it before myself, just looking it up, but It's possibly not working due to a defect in the C++ 98 standard.
So VC6 might not support it with the way it has been implemented.

Mauller commented:
Will have a go at this again tomorrow.

xezon commented:
The Ai generated workaround looks strange however.

```cpp
// Full specialization of helper for U = int
template <typename T>
struct MyClassImpl<T, int> {
```

It describes this as full specialization, but it still looks like partial. I expect this does not work.

Mauller commented:
> The Ai generated workaround looks strange however.
> 
> ```c++
> // Full specialization of helper for U = int
> template <typename T>
> struct MyClassImpl<T, int> {
> ```
> 
> It describes this as full specialization, but it still looks like partial. I expect this does not work.

I was thinking that we might have to just do
```cpp
#if defined(_msc_ver) && _msc_ver < 1300
Littlefiles = ntohl(Littlefiles)
#else
Littlefiles = betoh(Littlefiles)
#endif

In the win32 big file system then conditionally gate the endian compat to include winsock2 for vc6 and hide all the template code.

xezon reviewed (commented):
I think we can do it simpler by creating inline functions. This should be sufficient for non-enum types.

inline uint16_t betoh(uint16_t x) ...
inline uint32_t betoh(uint32_t x) ...

Mauller commented on Dependencies/Utility/Utility/endian_compat.h:
```diff
@@ -0,0 +1,198 @@
+
+#define htobe32(x) htonl(x)
+#define htole32(x) (x)
+#define be32toh(x) ntohl(x)
```
Yeah VC6 does not have any built in byte swapping functions so would have to roll our own.
i will conditionally exclude the templates if VC6 is used and add some overloaded functions like you mentioned above.

Mauller commented:
Hopefully this is the one lol.

Checked the byte swap macros by having them get compiled in VS22 and they worked fine.

Mauller commented:
Tested the VC6 builds and they work as expected.

xezon commented on Dependencies/Utility/Utility/endian_compat.h:
```diff
@@ -0,0 +1,209 @@
+__inline Short		betoh(Short value)		{ return be16toh(value); }
+__inline Int		betoh(Int value)		{ return be32toh(value); }
+
+// Partial template endian functions
```
What does this comment do?

Mauller replied:
It was mostly there just to help identify what the else is refering too, but can drop it or stick it inline with the else.

xezon replied:
I think it can be removed.

Mauller commented on Dependencies/Utility/Utility/endian_compat.h:
```diff
@@ -0,0 +1,209 @@
+__inline SwapType16 betoh(SwapType16 value) { return be16toh(value); }
+__inline SwapType32 betoh(SwapType32 value) { return be32toh(value); }
+__inline Short		betoh(Short value)		{ return be16toh(value); }
+__inline Int		betoh(Int value)		{ return be32toh(value); }
```
Had it in mind that VC6 didn't support standard ```inline``` can change this to normal ```inline``` call

Mauller commented on Dependencies/Utility/Utility/endian_compat.h:
```diff
@@ -0,0 +1,209 @@
+#if defined(_MSC_VER) && _MSC_VER < 1300
+typedef unsigned short SwapType16;
+typedef unsigned long SwapType32;
+typedef unsigned __int64 SwapType64;
```
Yeah thats fair if i add the int compat i can get rid of the conditional for this too since it was only there due to VC6 not supporting long long.

Mauller commented on Dependencies/Utility/Utility/endian_compat.h:
```diff
@@ -0,0 +1,209 @@
+
+#define htobe32(x) ( ((x & 0xFF000000) >> 24) | ((x & 0x00FF0000) >>  8) | ((x & 0x0000FF00) <<  8) | ((x & 0x000000FF) << 24) );
+#define htole32(x) (x)
+#define be32toh(x) ( ((x & 0xFF000000) >> 24) | ((x & 0x00FF0000) >>  8) | ((x & 0x0000FF00) <<  8) | ((x & 0x000000FF) << 24) );
```
nice spot.

Mauller commented on Dependencies/Utility/Utility/endian_compat.h:
```diff
@@ -0,0 +1,209 @@
+#define be16toh(x) ( (unsigned short)( ((x & 0xFF00) >>  8) | ((x & 0x00FF) <<  8) ) )
+#define le16toh(x) (x)
+
+#define htobe32(x) ( ((x & 0xFF000000) >> 24) | ((x & 0x00FF0000) >>  8) | ((x & 0x0000FF00) <<  8) | ((x & 0x000000FF) << 24) );
```
yeah i was going with the vertical cleanliness of the various functions around it. i might just break out the byte swap code seperately then refer to it in the blocks of endian macros.

Mauller commented on Dependencies/Utility/Utility/endian_compat.h:
```diff
@@ -0,0 +1,209 @@
+#else
+#define htobe16(x) ( (unsigned short)( ((x & 0xFF00) >>  8) | ((x & 0x00FF) <<  8) ) )
+#define htole16(x) (x)
+#define be16toh(x) ( (unsigned short)( ((x & 0xFF00) >>  8) | ((x & 0x00FF) <<  8) ) )
```
It doesn't really make a difference when you are using the value as a bitmask

Mauller commented:
Refactored taking the suggestions into account.

xezon commented on Dependencies/Utility/Utility/endian_compat.h:
```diff
@@ -0,0 +1,247 @@
+// VC6 compatible overloaded endian functions
+#if defined(_MSC_VER) && _MSC_VER < 1300
+
+// Big endian to hardware
```
Minor: I think h refers to host. See MacOS function name: OSSwapLittleToHostInt16

xezon reviewed (approved):
Looks good.

Mauller commented:
Pushed with tweaks to the comments, should be good now.

xezon commented:
This was quite the journey :P

Mauller commented:
> This was quite the journey :P

Always learning something along the way.
